### ISSUE TITLE
Possible conflict due to backup running during DST change? 

### ISSUE THREAD

**Issue by tomswartz07** [pgbackrest/pgbackrest#1226]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
- 2.27

2. PostgreSQL version:
- 10.14

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
- CloudFoundry PCF based on ubuntu-xenial

4. Did you install pgBackRest from source or from a package?
- From source (organization Build Team provided)

5. Please attach the following as applicable:
```
-------------------PROCESS START------------------- 
2020-11-04 14:32:19.683 P00 INFO: backup command begin 2.27: --compress --config=/var/vcap/jobs/standalone-postgresql/config/pgbackrest-s3.conf --lock-path=/var/vcap/store/pgbackrest --log-level-console=detail --log-level-file=detail --log-path=/var/vcap/sys/log/pgbackrest --pg1-path=/var/vcap/store/postgresql/data --process-max=4 --repo1-path=/crunchy/fb801dd3-f299-4668-8238-6607ad0e4905 --repo1-retention-diff=3 --repo1-retention-full=3 --repo1-s3-bucket=bucketnameredacted-1010221241 --repo1-s3-endpoint=s3-fips-us-gov-west-1.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-gov-west-1 --repo1-s3-uri-style=host --no-repo1-s3-verify-tls --repo1-type=s3 --no-resume --stanza=main --start-fast --stop-auto 
2020-11-04 14:32:21.331 P00 INFO: last backup label = 20201025-011518F_20201031-011512D, version = 2.27 
2020-11-04 14:32:21.331 P00 INFO: execute non-exclusive pg_start_backup(): backup begins after the requested immediate checkpoint completes 
2020-11-04 14:32:21.733 P00 INFO: backup start archive = 0000000100000004000000DB, lsn = 4/DB000028 
2020-11-04 14:32:23.341 P00 ERROR: [029]: new backup label '20201025-011518F_20201104-143222I' is not later than latest backup label '20201101-011519F' 
HINT: has the timezone changed? 
HINT: is there clock skew? 
2020-11-04 14:32:23.341 P00 DETAIL: socket statistics: objects 1, sessions 1, retries 0 
2020-11-04 14:32:23.341 P00 DETAIL: tls statistics: objects 1, sessions 1, retries 0 
2020-11-04 14:32:23.341 P00 INFO: http statistics: objects 1, sessions 1, requests 15, retries 0, closes 0 
2020-11-04 14:32:23.341 P00 INFO: backup command end: aborted with exception [029]
```

```
db (current) 
        wal archive min/max (10-1): 00000001000000040000005A/0000000100000004000000DA 

        full backup: 20201011-011519F 
                timestamp start/stop: 2020-10-11 01:15:19 / 2020-10-11 01:16:29 
                wal start/stop: 00000001000000040000005A / 00000001000000040000005A 
                database size: 54.9MB, backup size: 54.9MB 
                repository size: 6.8MB, repository backup size: 6.8MB 

        full backup: 20201018-011518F 
                timestamp start/stop: 2020-10-18 01:15:18 / 2020-10-18 01:16:25 
                wal start/stop: 000000010000000400000080 / 000000010000000400000080 
                database size: 55.9MB, backup size: 55.9MB 
                repository size: 7MB, repository backup size: 7MB 

        full backup: 20201025-011518F 
                timestamp start/stop: 2020-10-25 01:15:18 / 2020-10-25 01:16:58 
                wal start/stop: 0000000100000004000000A0 / 0000000100000004000000A0 
                database size: 56.9MB, backup size: 56.9MB 
                repository size: 7.1MB, repository backup size: 7.1MB 

        diff backup: 20201025-011518F_20201029-011515D 
                timestamp start/stop: 2020-10-29 01:15:15 / 2020-10-29 01:15:32 
                wal start/stop: 0000000100000004000000B3 / 0000000100000004000000B3 
                database size: 59.9MB, backup size: 21.3MB 
                repository size: 7.6MB, repository backup size: 3.1MB 
                backup reference list: 20201025-011518F 

        diff backup: 20201025-011518F_20201030-011514D 
                timestamp start/stop: 2020-10-30 01:15:14 / 2020-10-30 01:15:27 
                wal start/stop: 0000000100000004000000B7 / 0000000100000004000000B7 
                database size: 59.9MB, backup size: 21.3MB 
                repository size: 7.6MB, repository backup size: 3.1MB 
                backup reference list: 20201025-011518F 

        diff backup: 20201025-011518F_20201031-011512D 
                timestamp start/stop: 2020-10-31 01:15:12 / 2020-10-31 01:15:25 
                wal start/stop: 0000000100000004000000BB / 0000000100000004000000BB 
                database size: 59.9MB, backup size: 21.4MB 
                repository size: 7.6MB, repository backup size: 3.1MB 
                backup reference list: 20201025-011518F
```

7. Describe the issue:

Backups on these systems are regularly scheduled for 1am.
Full backup on Sunday 1am, all other days are diff type.

Following what appears to have been an error/issue on the 31st (when DST ends), no new `diff` backups can be created, resulting in the above error message.

Curiously, the backups referenced in the logs do not appear to be listed as valid in the pgbackrest info output. This is to say, backups which were supposed to have been performed on the days following 1 Nov are not listed in the info output, but they do have identifiers which are captured in the logs. 

It does appear that taking a new full backup will resolve this issue and allow pgbackrest to function as expected.

I am in the process of trying to obtain the logs from the 31st and 1st and will post here when they are available to me.

**Comment by tomswartz07** [comment]
Hi there:

Unfortunately, the nodes associated with this error were already recovered by the time we were able to obtain logs, so the latest available information is from today (5 Nov 2020)
I'm presently investigating if there are any nodes which still contain the log files from the 31st/1st.


However, I do have a copy of the `main-backup.log` which contains several instances where the backup label is noted to be "not later" than the previous.

It is attached here for review, along with the expire log.

[main-backup.log](https://github.com/pgbackrest/pgbackrest/files/5497626/main-backup.log)
[main-expire.log](https://github.com/pgbackrest/pgbackrest/files/5497644/main-expire.log)

**Comment by cmwshang** [comment]
Thanks, Tom. It sounds like the system is working again but we will investigate further given the information you provided.

**Comment by dwsteele** [comment]
Not really enough data here to say for sure, but it looks like `20201101-011519F` is somehow not getting into `backup.info` (even on reconstruct attempts), but it still exists on disk. This could be due to some kind of strange failure during the backup that we've never seen before.

As for this being DST related -- is DST enabled? The time used to format the backup labels comes from *PostgreSQL*, not the host. So it matters if PostgreSQL is setup with DST enabled. I can see how a backup started during the DST window could have an end time before the start time which might cause the backup to fail. If DST is not enabled this could be a coincidence, or a failure in some of part of the system that does have DST enabled.

According to the expire logs a backup with nearly the same label was expired:
```
remove expired backup 20201101-011512F
```
This indicates that two full backups were either taken within seconds of each other or one was taken after a DST change. I see a similar thing in the logs from 11/5 where backups are being taken in quick succession. That's pretty weird.

If the `20201101-011519F` still exists it would be good to get some info about the top level dir:
```
ls -lah /crunchy/fb801dd3-f299-4668-8238-6607ad0e4905/backup/main/20201101-011519F
```
The header from the `backup.manifest` or  `backup.manifest.copy` file would also be very helpful:
```
head -n 100  /crunchy/fb801dd3-f299-4668-8238-6607ad0e4905/backup/main/20201101-011519F/backup.manifest
```
or if that doesn't exist, then:
```
head -n 100  /crunchy/fb801dd3-f299-4668-8238-6607ad0e4905/backup/main/20201101-011519F/backup.manifest.copy
```
I wouldn't be surprised, though, if this backup is gone. If it didn't look valid enough to add to `backup.info` then expire probably would have removed it.

**Comment by tomswartz07** [comment]
Hey David:

Thanks for the analysis so far.

I'll try to obtain the info related to the `20201101-011519F` backup, as requested.


I should note that the PG settings have `UTC` timezone, but it is unclear what timezone is used for the underlying PCF stemcell (host OS).


>This indicates that two full backups were either taken within seconds of each other or one was taken after a DST change.

I have a feeling it might be the latter in this case.
There have been some documented issues with the underlying stemcell OS being used, and this would not be an unusual oversight or failure with regard to its operation, unfortunately.

I should note that the backup jobs are triggered by `cron` with the following schedule:
```
15 1 * * 7 vcap /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main --type=full backup
15 1 * * 1-6 vcap /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main --type=diff backup
```

Thanks again for the analysis so far, and I will update with any further information I can obtain.

**Comment by dwsteele** [comment]
I'm going to close this since there is no further information and I was not able to reproduce.

However, we do have a card (https://github.com/pgbackrest/pgbackrest/projects/2#card-42750170) to improve the way we are tracking backups in progress, which is the only thing I can think of that would have caused this issue. We'll update this issue when that change is made.

**Comment by tomswartz07** [comment]
Hi there;

Sorry for the delay; we were waiting for further updates from the customer and only just received new info now.

This error has occurred again in the automated environment.
I am still waiting for the contents of the backup manifest files, as per the previous discussion.

Here is the backrest info output:

```
stanza: main
   status: ok
   cipher: none

   db (current)
       wal archive min/max (10-1): 000000010000000200000020/00000001000000020000007A

       full backup: 20201108-011512F
           timestamp start/stop: 2020-11-08 01:15:12 / 2020-11-08 01:16:25
           wal start/stop: 000000010000000200000020 / 000000010000000200000020
           database size: 37.4MB, backup size: 37.4MB
           repository size: 4.4MB, repository backup size: 4.4MB

       full backup: 20201115-011511F
           timestamp start/stop: 2020-11-15 01:15:11 / 2020-11-15 01:16:56
           wal start/stop: 00000001000000020000003B / 00000001000000020000003B
           database size: 37.4MB, backup size: 37.4MB
           repository size: 4.4MB, repository backup size: 4.4MB

       full backup: 20201122-011516F
           timestamp start/stop: 2020-11-22 01:15:16 / 2020-11-22 01:16:20
           wal start/stop: 000000010000000200000054 / 000000010000000200000054
           database size: 37.4MB, backup size: 37.4MB
           repository size: 4.4MB, repository backup size: 4.4MB

       diff backup: 20201122-011516F_20201126-011511D
           timestamp start/stop: 2020-11-26 01:15:11 / 2020-11-26 01:15:21
           wal start/stop: 000000010000000200000061 / 000000010000000200000061
           database size: 37.4MB, backup size: 8.3KB
           repository size: 4.4MB, repository backup size: 490B
           backup reference list: 20201122-011516F

       diff backup: 20201122-011516F_20201127-011508D
           timestamp start/stop: 2020-11-27 01:15:08 / 2020-11-27 01:15:18
           wal start/stop: 000000010000000200000064 / 000000010000000200000064
           database size: 37.4MB, backup size: 8.3KB
           repository size: 4.4MB, repository backup size: 491B
           backup reference list: 20201122-011516F

       diff backup: 20201122-011516F_20201128-011508D
           timestamp start/stop: 2020-11-28 01:15:08 / 2020-11-28 01:15:19
           wal start/stop: 000000010000000200000068 / 000000010000000200000068
           database size: 37.4MB, backup size: 8.3KB
           repository size: 4.4MB, repository backup size: 491B
           backup reference list: 20201122-011516F
```

From the backup logs:

```
[.....]
2020-12-03 02:12:49.292 P00   INFO: incr backup size = 90.2KB
2020-12-03 02:12:49.292 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2020-12-03 02:12:49.392 P00   INFO: backup stop archive = 00000001000000020000007A, lsn = 2/7A0000F8
2020-12-03 02:12:49.395 P00 DETAIL: wrote 'backup_label' file returned from pg_stop_backup()
2020-12-03 02:12:49.397 P00   INFO: check archive for segment(s) 00000001000000020000007A:00000001000000020000007A
2020-12-03 02:12:49.531 P00   INFO: new backup label = 20201129-011501F_20201203-021247I
2020-12-03 02:12:49.581 P00   INFO: backup command end: completed successfully (2348ms)

-------------------PROCESS START-------------------
2020-12-03 02:12:49.705 P00   INFO: backup command begin 2.27: --compress --config=/var/vcap/jobs/standalone-postgresql/config/pgbackrest-s3.conf --lock-path=/var/vcap/store/pgbackrest --log-level-console=detail --log-level-file=detail --log-path=/var/vcap/sys/log/pgbackrest --pg1-path=/var/vcap/store/postgresql/data --process-max=4 --repo1-path=/crunchy/4953403c-5089-48d3-a4d1-978856eee5c1 --repo1-retention-diff=3 --repo1-retention-full=3 --repo1-s3-bucket=tile-backups-krstaging-20190326224449779500000006 --repo1-s3-endpoint=s3-fips-us-gov-west-1.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-gov-west-1 --repo1-s3-uri-style=host --no-repo1-s3-verify-tls --repo1-type=s3 --no-resume --stanza=main --start-fast --stop-auto
2020-12-03 02:12:51.328 P00   INFO: last backup label = 20201122-011516F_20201128-011508D, version = 2.27
2020-12-03 02:12:51.328 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the requested immediate checkpoint completes
2020-12-03 02:12:51.829 P00   INFO: backup start archive = 00000001000000020000007B, lsn = 2/7B000028
2020-12-03 02:12:53.148 P00  ERROR: [029]: new backup label '20201122-011516F_20201203-021252I' is not later than latest backup label '20201129-011512F'
                                    HINT: has the timezone changed?
                                    HINT: is there clock skew?
2020-12-03 02:12:53.148 P00 DETAIL: socket statistics: objects 1, sessions 3, retries 0
2020-12-03 02:12:53.148 P00 DETAIL: tls statistics: objects 1, sessions 3, retries 0
2020-12-03 02:12:53.148 P00   INFO: http statistics: objects 1, sessions 3, requests 15, retries 0, closes 2
2020-12-03 02:12:53.148 P00   INFO: backup command end: aborted with exception [029]
```

And expire logs:
```
-------------------PROCESS START-------------------
2020-12-03 02:12:49.581 P00   INFO: expire command begin 2.27: --config=/var/vcap/jobs/standalone-postgresql/config/pgbackrest-local.conf --log-level-console=detail --log-level-file=detail --log-path=/var/vcap/sys/log/pgbackrest --repo1-path=/var/vcap/store/pgbackrest --repo1-retention-diff=3 --repo1-retention-full=3 --stanza=main
2020-12-03 02:12:49.588 P00 DETAIL: archive retention on backup 20201115-011502F, archiveId = 10-1, start = 000000010000000200000039
2020-12-03 02:12:49.588 P00 DETAIL: no archive to remove, archiveId = 10-1
2020-12-03 02:12:49.688 P00   INFO: expire command end: completed successfully (107ms)
```

Please let me know if there is anything additional that would be useful to help diagnose this issue.

**Comment by cmwshang** [comment]
David had mentioned in his previous comment about getting some directory information (top level, backup.manifest, etc) - could you do that for this latest (and if not, maybe the next time it happens? What is puzzling is that the 20201129-011501F_20201203-021247I isn't showing up in the info command which makes me wonder if there is another repo somewhere.

**Comment by tomswartz07** [comment]
Hi Cynthia!

Thanks for the update. As it so happens, I finally got the info necessary from the customer this morning.


However, first, just a bit of context here:
Yes- there are two repositories for each of these clusters. One is locally stored (using `pgbackrest-local.conf`) and the other is stored on an S3-compatible device (Dell ECS, using `pgbackrest-s3.conf`).

The backup jobs are triggered via cron;
```
15 1 * * 1-6 /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main
15 1 * * 7 /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main
```

However, it appears that the issue is solely related to the S3 stored backups.


Attached are the manifest and manifest.copy files for what appears to be the latest backup in the S3 bucket, as well as the file listing of the storage directory.
[backupmanifest.tar.gz](https://github.com/pgbackrest/pgbackrest/files/5666713/backupmanifest.tar.gz)

**Comment by tomswartz07** [comment]
Following on from the previous message, I realize that perhaps the issue here is being introduced via the method that the backups are being performed.

Is it possible that this error is because of having two consecutive backup commands running as defined above?
If NTP updates the clock skew back a second or two, and the backups only take a second, I suppose its possible to have a backup 'start' before the previous one had finished, despite these being two separate repositories.

**Comment by cmwshang** [comment]
> Is it possible that this error is because of having two consecutive backup commands running as defined above?
> If NTP updates the clock skew back a second or two, and the backups only take a second, I suppose its possible to have a backup 'start' before the previous one had finished, despite these being two separate repositories.

Ahh, I'd suggest waiting for @dwsteele to chime in, but I would separate the backups in the cron for each repo by several minutes.

**Comment by dwsteele** [comment]
I'm not sure if delaying the backups will help since they are running against different repos.

What it looks like is `20201129-011512F` is on disk but not in `backup.info`, perhaps because it does not look quite valid.

I still don't think we are getting a consistent set of data here. For instance the screenshot of the S3 bucket provided shows diff backups but no incrementals -- even though that's what the logs shows. Are all incr backups failing? How are these diff backups being run.

These cron jobs do not look valid and do not match up with the logs:
```
15 1 * * 1-6 /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main
15 1 * * 7 /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main
```
They do not specify the command to run or the backup type. Were they redacted?

Also using && here is not a good idea. If the first backup fails then the second one won't run.

If there are any logs for the `20201129-011512F` backup we would definitely like to see those. I still suspect this backup is error'ing out somehow.

**Comment by tomswartz07** [comment]
Hi David;

Thanks for the update.

>[…] since they are running against different repos.

This was my suspicion as well, but I thought it was worth highlighting the hypothesis. Thanks for confirming.

>They do not specify the command to run or the backup type. Were they redacted?

Apologies- there was a transposition error on my part while taking the 'template' for the config and filling in the relevant details for the issue here.

The corrected cron template is as follows;
```
15 1 * * 7  /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main --type=full backup && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main --type=full backup
15 1 * * 1-6  /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-local.conf --stanza=main --type=diff backup && /usr/bin/pgbackrest --config=/var/vcap/store/pgbackrest/config/pgbackrest-s3.conf --stanza=main --type=diff backup
```

Unfortunately, the only logs we have are the ones provided in the previous message, [here](https://github.com/pgbackrest/pgbackrest/files/5666713/backupmanifest.tar.gz).

This error was preventing their upgrade process, so they had pulled the files requested at the time and after performing a manual full backup were able to continue with the upgrade.
Presumably, following the full backup the offending backup set was expired.

**Comment by dwsteele** [comment]
This issue should be resolved by #1272. It was not related to the DST change.
